src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+
+              startPeriodicAllocation += spInterval;
+              counter++;
+
+              // update remaining DTI time for consistency
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
+            }
+
+          AddAllocationPeriod (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
+                               sourceAid, info.GetDestinationAid (),
+                               startFirstAllocation, allocDuration, spInterval, blocks);
+          status.SetSuccess ();
+        }
+      else
+        {
+          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected.
+          status.SetFailure ();
+        }
+
+    }
+  else
+    {
+
+      uint32_t endAlloc;
+      uint32_t slotDur;
+      for (auto it = m_availableSlots.begin (); it != m_availableSlots.end (); ++it)
+        {
+          slotDur = it->second - it->first;
+
+          if (slotDur > allocDuration)
+            {
+
+              endAlloc = AllocateSingleContiguousBlock (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
+                                                        sourceAid, info.GetDestinationAid (), it->first, allocDuration);
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
+
+              // The following function modifies m_availableSlots, on which this
+              // loop is iterating. Commonly, it is a bad practice but, as we break 
+              // the loop if we enter this condition, the damage is avoided.
+              UpdateAvailableSlots (it->first, endAlloc);
+              status.SetSuccess ();
+              break;
+            }
+          else
+            {
+              status.SetFailure ();
+            }
+        }
+
+    }
+
+  return status;
+}
+
+uint8_t
+PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t spInterval)
+{
+  NS_LOG_FUNCTION (this);
+
+  auto it = m_availableSlots.begin ();
+  uint32_t startAlloc = it->first;
+  uint32_t slotDuration = 0;
+  uint8_t blocks = 0;
+
+  while (it != m_availableSlots.end ())
+    {
+      if (startAlloc < it->first)
+        {
+          // Periodicity has been broken, we stop the algorithm.
+          break;
+        }
+
+      slotDuration = it->second - startAlloc;
+
+      if (allocDuration + m_guardTime > slotDuration)
+        {
+          if (blocks == 0)
+            {
+              // go on until eventually find the first available slot that fits this SP
+              // note that this also cover the condition where no slot satisfies the requirement
+              ++it;
+              startAlloc = it->first;
+              continue;
+            }
+          else
+            {
+              // if one or more periodic SPs already allocated, now the periodicity
+              // is broken and the algorithm stops
+              break;
+            }
+        }
+
+      blocks++;
+      startAlloc += spInterval;
+
```

Comment:
Only an octet is reserved for Number of Blocks, avoid overflows when BI duration is extremely long or the periodicity is extremely short

```suggestion
      if (blocks == 255)
        {
          // Number of blocks described by an octet: only up to 255 blocks
          break;
        }
```